### ISSUE TITLE
When Request Method is "HEAD" and setFollowRedirect is set to true. The method of the second redirect request is GET.

### ISSUE THREAD

**Issue by lakxtxue** [AsyncHttpClient/async-http-client#1728]
When Request Method is "HEAD" and setFollowRedirect is set to true. The method of the second redirect request is GET.
When I download a large file, it will time out here, in fact it has been downloading

`Caused by: java.util.concurrent.TimeoutException: Request timeout to 127.0.0.1/127.0.0.1:8888 after 60000 ms
	at org.asynchttpclient.netty.timeout.TimeoutTimerTask.expire(TimeoutTimerTask.java:43)
	at org.asynchttpclient.netty.timeout.RequestTimeoutTimerTask.run(RequestTimeoutTimerTask.java:50)
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:672)
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:747)
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:472)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)`

**Comment by tomgs** [comment]
@lakxtxue hi - I've applied to maintain this library, so I'm taking a look at open issues to see whether I can maybe help out.

As a general note, a paste of the code tends to help in these issues - if not for better comprehension then at least to save a few lines of reproduction:) 

Just to make sure I follow: you've got a large resource located at `urlA`. When dispatching a `HEAD` request to `urlA`, you're getting back a 301 that points to `urlB`, which is followed because `setFollowRedirect` is set to `true`. The request to `urlB` is not a `HEAD` but a `GET`.

Since the resource hiding behind `urlA` (and in effect, `urlB`) is quite large, you're getting a timeout instead of getting a response back from `urlB`.

What I assume you'd like to happen is for the same request type that was originally dispatched to be used in the followup requests as well.

Is this a correct assessment?

**Comment by lakxtxue** [comment]
@tomgs hi, yes, that's what I mean, because I use http to download large files, I need to get the size of the file first. When accessing this file resource, there will be a redirection to the real address. I use the **HEAD** request method to get the file size. **AsyncHttpClient** encounters a **301** When it was directed, it did not continue to access with a **HEAD** request but used a **GET** request to access the redirected address. Now I use **HttpURLConnection** instead, and I get the result I want correctly.
This is my code using **HttpURLConnection**
```
URL u = new URL(url);
HttpURLConnection httpConnection = (HttpURLConnection) u.openConnection();
httpConnection.setRequestMethod("HEAD");
httpConnection.connect();
this.dataSize = httpConnection.getContentLengthLong();
```

I used AsyncHttpClient to do the same thing. I also set the **HEAD** request method and configured **setFollowRedirect** to **true**. The result was a timeout. I found through packet capture that AsyncHttpClient used **GET** request when redirecting and started to download files. I hope it will behave like **HttpURLConnection**.

I don’t know if the expression is clear or not, because my English is not very good. I have found a lot of http libraries. AsyncHttpClient is the only one that can provide me with the greatest flexibility. I really hope it will get better and better.

**Comment by TomGranot** [comment]
@lakxtxue I see what you're saying, and it's perfectly clear. Your English is much better than many of the non-native English speakers I've met (I'm also not a native speaker, btw!).

I actually wasn't sure what the HTTP protocol specification says about method handling in redirects.

I didn't go to the spec itself, but [MDN says](https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections#Permanent_redirections) that when a user-agent (like AsyncHttpClient) sends a request and encounters a 301, only `GET` requests remain unchanged, an the rest of the HTTP methods "may or may not be changed".

I am still familiarising myself with the library, but an [initial search](https://github.com/AsyncHttpClient/async-http-client/search?q=setFollowRedirect&type=Commits) seems to show that [it's related](https://github.com/AsyncHttpClient/async-http-client/commit/fbf4fdec166f736146c9ca9a562b395089dc2901) to underlying [Netty provider](https://github.com/netty/netty) this library is using. 

I'd say that it's reasonable for the method to persist between redirects, but would have to dig deeper later tonight to see how it can be actually implemented and open a PR. If I forget to follow up please ping me in a few days!

**Comment by TomGranot** [comment]
Alright, so I think I found the culprit:

https://github.com/TomGranot/async-http-client/blob/master/client/src/main/java/org/asynchttpclient/netty/handler/intercept/Redirect30xInterceptor.java#L88-L91

**Comment by TomGranot** [comment]
So I asked around and received other use cases where it makes sense to have a the next request in the cycle be a `GET` instead. I think this can be easily solved by adding another config option for the `RequestBuilder` called `persistMethodBetweenRedirects` that will basically allow you to do `HEAD` all the way if you wanted to.

I'll write a PR in a bit.

**Comment by slandelle** [comment]
Here’s what I do in Gatling: https://github.com/gatling/gatling/commit/a87729b7d1ce503ea16184b68b6b436d50eab658

**Comment by TomGranot** [comment]
Ah, that makes sense, this way you allow it for `HEAD` and `OPTIONS` but disallow it for `POST` that can pass too much information away. Thanks @slandelle!

**Comment by TomGranot** [comment]
Still need to open a PR, just not getting to it. Will do so sometime this week (need to re-write the tests so Maven won't be angry).

**Comment by TomGranot** [comment]
Keeping this here for now, will move to the pr a bit later - seems that `AsyncStreamHandlerTest.asyncOptionsTest()` is flaky - `http://apache.org` returns either 4 or 5 methods sporadically:

```
tom@Toms-MacBook-Pro async-http-client-fork % curl -i -X OPTIONS http://www.apache.org/
HTTP/1.1 200 OK
Date: Mon, 28 Sep 2020 11:24:45 GMT
Server: Apache/2.4.18 (Ubuntu)
Allow: POST,OPTIONS,GET,HEAD
Cache-Control: max-age=3600
Expires: Mon, 28 Sep 2020 12:24:45 GMT
Content-Length: 0
Content-Type: text/html

tom@Toms-MacBook-Pro async-http-client-fork % curl -i -X OPTIONS http://www.apache.org/
HTTP/1.1 200 OK
Date: Mon, 28 Sep 2020 11:53:21 GMT
Server: Apache
Allow: GET,POST,OPTIONS,HEAD,TRACE
Cache-Control: max-age=3600
Expires: Mon, 28 Sep 2020 12:53:21 GMT
Content-Length: 0
Content-Type: text/html
```

**Comment by TomGranot** [comment]
One of the really fun things about long-standing repos is that each piece of code has a rich history behind it, with various people facing the same(-ish) problem at different points in time, adding fixes and thought-nuggets into the code over the years. This seems to be exactly one of those cases - looking at https://github.com/TomGranot/async-http-client/commits/master/client/src/test/java/org/asynchttpclient/Head302Test.java reveals that this test actually has been fixed in the past.

Specifically in our case, it seems that https://github.com/TomGranot/async-http-client/commit/3c25a42289bf679cfff40ca5ef0e77be85215deb#diff-64fc1f6f6a5d572c250342b4cdfda122 made this very test added the behaviour that causes this endless loop we're looking at right now - the [`Head302Hander`](https://github.com/TomGranot/async-http-client/blob/master/client/src/test/java/org/asynchttpclient/Head302Test.java#L70-L84) is specifically written to keep on appending `_moved` to the URI and send back 302 after every 302 response.

In plain terms, this means that the test expects a `HEAD` to be followed with a `GET` and then terminate successfully. Since we now send followup `HEAD`s upon a `HEAD` where `setFollowRedirect` is true, we should probably fix it in the handler - decide on a number of redirects following which a GET is sent instead of a HEAD in this redirect chain, which is less than the maximum number of allowed redirects (defaults to 5, as the test dump above shows).

**Comment by TomGranot** [comment]
@lakxtxue Merged to master, should be deployed in 2.12.3.
